### PR TITLE
[Fix][Connector-V2] Fixed clickhouse connectors cannot stop under multiple parallelism

### LINKED ISSUE #7897: [Bug] [clickhouse] CK连接器任务抽取完毕后无法正常停止
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/seatunnel/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.

### What happened

当我通过clickhouse连接器抽取数据的时候，觉得速度比较慢，想调整并行度（虽然看到官网提示clickhouse连接器目前不支持并行），把parallelism 调整为2，这个时候当任务抽取完毕的时候，该任务并不会停止
已经验证过，必现，当我把并行度改为1的时候就没问题

虽然不支持并行度，但是期望调整 parallelism 参数并不会影响任务的正常结束
另外期望能开发该连接器，使其支持并行
可以提供思路，我来改

### SeaTunnel Version

2.3.7

### SeaTunnel Config

```conf
env {
  parallelism = 2
  job.mode = "BATCH"
}


source{
    Clickhouse {
    host = ""
    database = "fdc"
    sql = "select * from fdc.trace_data WHERE traceTime > '2024-10-24 04:00:00' and traceTime < '2024-10-24 04:00:20'"
    username = "reader"
    password = "reader@123"
    server_time_zone = "UTC"
    result_table_name = "fdc.trace_data"
    clickhouse.config = {
	  "max_memory_usage": "128000000000",
	  "max_bytes_before_external_group_by": "50000000000",
      "socket_timeout": "300000"
    }
  }
}


sink {
    jdbc {
        url = "jdbc:oceanbase:///ods"
        driver = "com.oceanbase.jdbc.Driver"
        user = "root"
        password = "Bh%Yac{C9l-~#?~dq[}bSQO5ng8Rr?w"
        compatible_mode = "mysql"
        generate_sink_sql = true
        database = ods
        table = trace_data
        data_save_mode = APPEND_DATA
    }
}
```

### Running Command

```shell
./bin/seatunnel.sh --config job/ck2ob.config -e local
```

### Error Exception

```log
任务抽取完毕，但是停不下来
```

### Zeta or Flink or Spark Version

zeta

### Java or Scala Version

java

### Screenshots

119180a7b9e22a7afeadc960f80a42a

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by YOMO-Lee**
Fixed the bug where clickhouse connectors cannot stop properly under multiple parallelism

### CODE REVIEW

File: seatunnel-engine/seatunnel-engine-common/src/main/resources/seatunnel.yaml
Comment (corgy-w) [inline seatunnel-engine/seatunnel-engine-common/src/main/resources/seatunnel.yaml:23]: Default value of `print-job-metrics-info-interval` is 60. This is valid without change

Comment (YOMO-Lee) [reply seatunnel-engine/seatunnel-engine-common/src/main/resources/seatunnel.yaml:23]: I know that the default value is also used if it is not changed here, but here is a parameter to let you know where you can change it.
Otherwise, if you run an example in the project and want to change the parameters, the results will not work

Comment (YOMO-Lee) [reply seatunnel-engine/seatunnel-engine-common/src/main/resources/seatunnel.yaml:23]: > Default value of `print-job-metrics-info-interval` is 60. This is valid without change

I think this parameter can be added here, at least adding the default value will not cause any other bad effects
But it worked for me, at first I wanted to change the configuration, I didn't know where to change, I searched a lot of places in the project, and it didn't take effect after the change, and then it was someone else who pointed out that I knew to change here

Comment (corgy-w) [reply seatunnel-engine/seatunnel-engine-common/src/main/resources/seatunnel.yaml:23]: get

File: seatunnel-e2e/seatunnel-connector-v2-e2e/connector-clickhouse-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/ClickhouseIT.java
Comment (Hisoka-X) [inline seatunnel-e2e/seatunnel-connector-v2-e2e/connector-clickhouse-e2e/src/test/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/ClickhouseIT.java]:
```suggestion
```

File: seatunnel-engine/seatunnel-engine-common/src/main/resources/seatunnel.yaml
Comment (Hisoka-X) [reply seatunnel-engine/seatunnel-engine-common/src/main/resources/seatunnel.yaml:23]: You can configure it in your local environment. But do not commit it. Please revert it.

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java]:
```suggestion
        if (assigned < 0) {
            assigned = subtaskId;
            context.assignSplit(subtaskId, new ClickhouseSourceSplit());
            context.signalNoMoreSplits(subtaskId);
        } else {
            context.signalNoMoreSplits(subtaskId);
        }
```

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java
Comment (Hisoka-X) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java]:
```suggestion
            this.splits.clear();
        }
        if (noMoreSplit
                && splits.isEmpty()
                && Boundedness.BOUNDED.equals(readerContext.getBoundedness())) {
            log.info("Closed the bounded ClickHouse source");
            this.readerContext.signalNoMoreElement();
            this.splits.clear();
        }
```

**Review by Hisoka-X — COMMENTED** [review]
Please add checkpoint lock when read spilt. Please refer https://github.com/apache/seatunnel/blob/50113e77cf6729c833bca6a5e20ef397b3992782/seatunnel-connectors-v2/connector-doris/src/main/java/org/apache/seatunnel/connectors/doris/source/reader/DorisSourceReader.java#L76

**Comment by Hisoka-X** [comment]
cc @zhilinli123 as well.

**Comment by zhilinli123** [comment]
> cc @zhilinli123 as well.
1.Thanks for ping me, I think pom needs to be added 
```

   <dependency>
            <groupId>org.apache.seatunnel</groupId>
            <artifactId>seatunnel-api</artifactId>
            <version>${project.version}</version>
        </dependency>
```

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java
Comment (zhilinli123) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java]: >  Write it like this?
 
```

        if (assigned < 0) {
            assigned = subtaskId;
            context.assignSplit(subtaskId, new ClickhouseSourceSplit());
        }
        context.signalNoMoreSplits(subtaskId);

```

**Comment by YOMO-Lee** [comment]
> > cc @zhilinli123 as well.
> > 1.Thanks for ping me, I think pom needs to be added
> 
> ```
> 
>    <dependency>
>             <groupId>org.apache.seatunnel</groupId>
>             <artifactId>seatunnel-api</artifactId>
>             <version>${project.version}</version>
>         </dependency>
> ```

I don't quite understand where this should be added and where it needs to be used. Currently, I haven't found any places where it needs to be used

**Comment by zhilinli123** [comment]
> > > cc @zhilinli123 as well.
> > > 1.Thanks for ping me, I think pom needs to be added
> > 
> > 
> > ```
> > 
> >    <dependency>
> >             <groupId>org.apache.seatunnel</groupId>
> >             <artifactId>seatunnel-api</artifactId>
> >             <version>${project.version}</version>
> >         </dependency>
> > ```
> 
> I don't quite understand where this should be added and where it needs to be used. Currently, I haven't found any places where it needs to be used

Boundedness.BOUNDED  quoted https://github.com/YOMO-Lee/seatunnel/blob/clickhouse/seatunnel-connectors-v2/connector-clickhouse/pom.xml

**Comment by zhilinli123** [comment]
Waiting for the start of the current pr merger modify this code: https://github.com/apache/seatunnel/pull/7529

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java
Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java]: > > Write it like this?
> 
> ```
> 
>         if (assigned < 0) {
>             assigned = subtaskId;
>             context.assignSplit(subtaskId, new ClickhouseSourceSplit());
>         }
>         context.signalNoMoreSplits(subtaskId);
> ```

+1

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java
Comment (zhilinli123) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java]: `Boundedness.BOUNDED.equals(readerContext.getBoundedness())`
This can be removed I think because clickhouse only supports batch mode reading
cc @Hisoka-X

Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java]: Keep it for now. The purpose of keeping it is to unify the processing logic of most connectors, to facilitate reference for other developers and reduce the chance of bugs.

Comment (zhilinli123) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java]: > Keep it for now. The purpose of keeping it is to unify the processing logic of most connectors, to facilitate reference for other developers and reduce the chance of bugs.

What if the user uses processing mode to write STREAMING? It's unnecessary code and I think it's better to get rid of it Developers adding features are certainly familiar with their read patterns

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java
Comment (zhilinli123) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java:104]: Extract clickhouse Close the task method
```
                this.readerContext.signalNoMoreElement();
                this.splits.clear();
```

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java
Comment (zhilinli123) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java]: ditto

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java
Comment (zhilinli123) [inline seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java]:
```

        if (assigned < 0) {
            assigned = subtaskId;
            context.assignSplit(subtaskId, new ClickhouseSourceSplit());
        }
        context.signalNoMoreSplits(subtaskId);
```

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java
Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java]: > What if the user uses processing mode to write STREAMING?

Sorry, I don get it. Could you share more details? You meaning set `job.mode` to `STREAMING`?

Comment (zhilinli123) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java]: > > What if the user uses processing mode to write STREAMING?
> 
> Sorry, I don get it. Could you share more details? You meaning set `job.mode` to `STREAMING`?
```

env {
  job.mode = "STREAMING"
  parallelism = 2
}
source {
  # This is a example source plugin **only for test and demonstrate the feature source plugin**
  Clickhouse {
    host = "127.0.0.1:8123"
    database = "default"
    sql = "select * from students"
    username = "default"
    password = ""
  }
  # If you would like to get more information about how to configure seatunnel and see full list of source plugins,
  # please go to https://seatunnel.apache.org/docs/connector-v2/source/ClickhouseSource
}


sink {
  Console {
  }
  # If you would like to get more information about how to configure seatunnel and see full list of sink plugins,
  # please go to https://seatunnel.apache.org/docs/connector-v2/sink
}

```

So for example in this case the user will never exit the program, essentially no matter how you configure the read mode clickhouse only supports batch mode, right

Comment (Hisoka-X) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java]: If users set to STREAMING, but source only supported BATCH mode. The job will execute as BATCH mode. It depends on https://github.com/apache/seatunnel/blob/a8d0d4ce77a1ac9116f93381dfe296d22d8f0e8b/seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSource.java#L160

Comment (zhilinli123) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java]: Understood, but an extra judgment is a redundant operation

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java
Comment (YOMO-Lee) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceSplitEnumerator.java]: Okay, I didn't notice this

File: seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java
Comment (YOMO-Lee) [reply seatunnel-connectors-v2/connector-clickhouse/src/main/java/org/apache/seatunnel/connectors/seatunnel/clickhouse/source/ClickhouseSourceReader.java]: OK

**Comment by zhilinli123** [comment]
Wait for CI pass

**Review by Hisoka-X — APPROVED** [review]
LGTM if ci passes. Thanks @YOMO-Lee and @zhilinli123 @corgy-w for review!

**Comment by zhilinli123** [comment]
> LGTM if ci passes. Thanks @YOMO-Lee and @zhilinli123 @corgy-w for review!

+1

**Review by zhilinli123 — APPROVED** [review]

**Comment by hailin0** [comment]
waiting for ci passed

**Review by hailin0 — APPROVED** [review]

**Pull request merged**
